apps/confidential/confidential.go

```diff
@@ -0,0 +1,364 @@
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/cache"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/internal/client"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/internal/requests"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/internal/requests/ops/accesstokens"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/internal/requests/ops/authority"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/internal/shared"
+)
+
+/*
+Design note:
+
+confidential.Client uses client.Base as an embedded type. client.Base statically assigns its attributes
+during creation. As it doesn't have any pointers in it, anything borrowed from it, such as
+Base.AuthParams is a copy that is free to be manipulated here.
+
+Duplicate Calls shared between public.Client and this package:
+There is some duplicate call options provided here that are the same as in public.Client . This
+is a design choices. Go proverb(https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s):
+"a little copying is better than a little dependency". Yes, we could have another package with
+shared options (fail).  That divides like 2 options from all others which makes the user look
+through more docs.  We can have all clients in one package, but I think separate packages
+here makes for better naming (public.Client vs client.PublicClient).  So I chose a little
+duplication.
+
+.Net People, Take note on X509:
+This uses x509.Certificates and private keys. x509 does not store private keys. .Net
+has some x509.Certificate2 thing that has private keys, but that is just some bullcrap that .Net
+added, it doesn't exist in real life.  Seriously, "x509.Certificate2", bahahahaha.  As such I've
+put a PEM decoder into here.
+*/
+
+// TODO(msal): This should have example code for each method on client using Go's example doc framework.
+// base usage details should be includee in the package documentation.
+
+// AuthenticationResult contains the results of one token acquisition operation.
+// For details see https://aka.ms/msal-net-authenticationresult
+type AuthenticationResult = client.AuthenticationResult
+
+type Account = shared.Account
+
+// CertFromPEM converts a PEM file (.pem or .key) for use with NewCredFromCert(). The file
+// must have the public certificate and the private key encoded. The private key must be encoded
+// in PKCS8 (not PKCS1). This is usally denoted by the section "PRIVATE KEY" (instead of PKCS1's
+// "RSA PRIVATE KEY"). If a PEM block is encoded and password is not an empty string, it attempts
+// to decrypt the PEM blocks using the password. This will return multiple x509 certificates,
+// though this use case should have a single cert. Multiple certs are due to certificate
+// chaining for use cases like TLS that sign from root to leaf.
+func CertFromPEM(pemData []byte, password string) ([]*x509.Certificate, crypto.PrivateKey, error) {
```

Comment:
Is returning the certs from the intermediate authorities actually valuable?  How do I know which cert in the slice is the one I use for authentication (you can't authenticate with a cert from an intermediate authority, we had this bug in our current authentication library).

Reply:
Well, in this context I think there is only the one certificate (however, I can't be SURE, so I am handling them all).  But.....  in a TLS world, you need all of them or you can't validate the chain and it won't work.

Case in point, when I had to write an internal keyvault client to do TLS for Go, I had to figure this all out.  And to be able to serve TLS, you needed the entire chain that if I remember correctly goes into some type of pool.  I'd have to look back and the code for details.

Also Microsoft issued certs will not natively work with Go's PKCS12 package, I had to do a fork to get it to work.

Reply:
I'll leave further convesation into that bug.

Reply:
There can definitely be more than one certificate.  See https://github.com/Azure/azure-sdk-for-go/issues/5412 for details on the problems as well as our solution.

My problem here is that this func can return multiple certificates however I don't know which one to pass to `confidential.NewCredFromCert()`.  If we have a legitimate use-case for returning all the certs I'm fine with that however we must document which cert is the one to use to authenticate (ideally the first in the slice).

Reply:
So, I don't think that fix is valid, if you are referring to:
```go 
  if _, ok := block.Headers["localKeyId"]; ok {
              cert, err = x509.ParseCertificate(block.Bytes)
              if err !=  nil {
                  fmt.Printf("Error parsing certificate from pem block:%v\n", err)
                  continue
              }
            }
```
I can find no option header standard for PEM that includes "locakKeyId" to tell you that something is the leaf certificate.  

When I look at this fix, it seems that the origin is from PKCS12.  PKCS12 is a different standard.  When a specific library does a PKCS12 to PEM conversion, it pops this in as an optional header.  

This *looks* to be coming from an asn1 id: asn1.ObjectIdentifier([]int{1, 2, 840, 113549, 1, 9, 21}) stored in PKCS12. (source; https://gitlab.cncf.ci/kubernetes/kubernetes/blob/a610ffd71ac8318c1cb2ddc4a234350e61162575/vendor/golang.org/x/crypto/pkcs12/pkcs12.go)

In standard PEM, this is not available unless a conversion was made from this specific pkcs12 package from what I see.

I don't think that fix is a valid fix for this or the library you linked.

So I think your question is actually still open (for both packages), how do you identify the leaf cert in a PEM encoding for all cases.  At this point I don't know that answer.

Reply:
The fix was to find the certificate with the same public key as that of the provided private key, see https://github.com/Azure/go-autorest/blob/master/autorest/adal/persist.go#L111.

Reply:
Interesting, though that is with a PKCS1 key instead of a PKCS8 key (which is what I added support for).  I'll have to look back, I though the standard for PKCS1 keys was to use RSA PRIVATE KEY and not PRIVATE KEY.  

Though, PEM seems to be very loose. 

Maybe I really need to add support for PKCS1 and PKCS8 and try to decode both types regardless of what the text is.  

I can't imagine how many iterations of features causes certificate/key standards to be such a mess.  ASN1.DER/CER, sometimes base64, sometimes not, inside PKCS1, PKCS12, PKCS8, ............

Text headers with options not listed......

Reply:
Yes agreed, it's quite the quagmire.

What if for now we drop support for certs with intermediate authorities and add it later?  As it stands now, this is pretty much not usable unless we assume we get one cert back (i.e. always use the first cert in the slice) which seems brittle.

Reply:
Alternative is keeping this (perhaps with a few additions) and doing testing with common scenarios where this is needed. We can fail with some meaningful message of we detect something is not working as expected.

Reply:
I'm going to defer this change to the next CL.  This will get the big changes out. 

The follow-up for this is on my next change.

Reply:
https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/122